eclipse-theia/theia
# Implement internationalization support

#### What it does

Closes #1327

Adds internationalization support for Theia. The following features are supported:

- [x] Support for [vscode language packs](https://code.visualstudio.com/docs/getstarted/locales) to localize Theia itself.
- [x] Support for localized vscode extensions (both `package.json` and runtime translations via `nls.localize()`)
- [x] Support for custom Theia localizations through a `LocalizationContribution` interface.
- [x] Support for localization of the monaco editor (search widget, context menu and so on).
- [x] Showing the original text under the localized value (for commands).

Things that are out of scope for this PR (and will probably be addressed in another PR):

- Support for vscode language packs to localize other vscode extensions (like the builtins)

#### How it works

1. The `core/browser` package contains a `nls` namespace with a `localize` function similiar to vscode. It accepts a translation key, a default value (usually the original english value) and additional args. The translations are loaded before any other imports so that they are available during load-time of any other code.
2. The backend application loads vscode language packs and Theia's `LocalizationContribution`s and places them in a `LocalizationProvider`. Additionally, the backend exposes a `/i18n/<locale>` HTTP endpoint to retrieve localizations for any locale.
3. The vscode plugin host starts the vscode extension process with the locale that is currently used by the user. This allows vscode extensions to translate themselves.
4. The `monaco-loader.ts` uses the frontend locale as well to configure monaco to use the currently active language.

#### How to test

0. Test that the `Configure display language` command can be invoked but only displays `en` (even though german translations exist in the backend, but no language pack is installed)
1. Download & install the german language pack (v1.50, search for `german` in the Theia extension menu)
2. Download & install [the i18n sample extension in german](https://github.com/eclipse-theia/theia/files/6658579/i18n-sample-0.1.0.zip)
3. Execute the `Hello` and `Bye` command contributed by the i18n sample extension. Those will simply show `Hello` and `Bye` respectively in a notification.
4. Execute the `Configure Display Language` command and choose the german locale `de`.
5. The application will be reloaded (you will maybe have to reload it yourself) and partially translated into german. You will see this in the main menu as well as the workspace and some common commands. *(Test for vscode localization packs)*
6. Execute the `Hallo` (Hello) and `Auf Wiedersehen` (Bye) command, which will now display german text as a notification. *(Test for vscode extensions to correctly localize themselves and for Theia to read the translated `package.json` correctly)*
7. In the context menu of the file explorer, you will see that the `Download` menu item has been replaced with `Herunterladen`. This was done with a german `LocalizationContribution`.
8. (optionally) Replace the `editor.main.nls.de.js` in the `examples/browser/lib/vs/editor` source directory with [this file](https://github.com/eclipse-theia/theia/files/6577192/editor.main.nls.de.zip). Run `gzip -k editor.main.nls.de.js` in that directory. Reloading the Theia application and using `de` locale will now show german translation in the monaco search widget (`Ctrl+F`) as well as in the context menu. *This is a workaround, see below for more information.*

#### Open Issues

* Do we need internationalization support for numbers and dates? - _Will be addressed in a separate PR, if at all_
* How about right to left languages?
* The `@theia/monaco-editor-core` package contains only english translations. Is this done on purpose? Replacing the `nls.*.json` files with the correct translations will translate monaco correctly.
* The translations for the monaco editor rely on the order of certain language pack versions. This isn't an issue for the default languages (like `de`, `it`, `fr`...), but makes it virtually impossible for vscode language packs to actually contribute to monaco's translation process.
* Language Packs can't be uninstalled without restarting the server (all localizations are held in memory after startup), as we don't know which translations are contributed by vscode language packs and which by Theia extensions **yet**.
* Currently the `LocalizationProvider` holds a single active language for the whole application. Is there a way to hold a value for every user?
* How do we handle translations contributed by Theia extensions without having a language pack installed? Selecting the german locale e.g. will just translate Theia's own commands, not the rest.
* Theia's quick-input-bar doesn't filter by the details of the item, unlike vscode. This makes finding commands quite difficult after switching the locale.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
